crates/oxc_semantic/src/const_enum.rs

```diff
@@ -0,0 +1,263 @@
+#[derive(Debug, Default, Clone)]
+pub struct ConstEnumTable {
+    /// Map of const enum symbol IDs to their information
+    pub enums: FxHashMap<SymbolId, NormalizedConstEnumInfo>,
+}
+
+impl ConstEnumTable {
+    /// Create a new const enum table
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add a const enum to the table
+    pub fn add_enum(&mut self, symbol_id: SymbolId, enum_info: NormalizedConstEnumInfo) {
+        self.enums.insert(symbol_id, enum_info);
+    }
+
+    /// Get all const enums
+    pub fn enums(&self) -> impl Iterator<Item = (&SymbolId, &NormalizedConstEnumInfo)> {
+        self.enums.iter()
+    }
+}
+
+pub struct ConstantEnumCtx<'b, 'a: 'b> {
+    const_enum_members: &'b FxHashMap<SymbolId, ConstEnumMemberInfo<'a>>,
+    scoping: &'a Scoping,
+    builder: oxc_ast::AstBuilder<'a>,
+}
+
+impl<'b, 'a> ConstantEnumCtx<'b, 'a> {
+    pub fn new(
+        const_enum_members: &'b FxHashMap<SymbolId, ConstEnumMemberInfo<'a>>,
+        scoping: &'a Scoping,
+        builder: oxc_ast::AstBuilder<'a>,
+    ) -> Self {
+        Self { const_enum_members, scoping, builder }
+    }
+}
+
+impl<'a> GlobalContext<'a> for ConstantEnumCtx<'_, 'a> {
+    fn is_global_reference(&self, ident: &IdentifierReference<'a>) -> bool {
+        ident
+            .reference_id
+            .get()
+            .and_then(|reference_id| {
+                let reference = self.scoping.references.get(reference_id)?;
+                let symbol_id = reference.symbol_id()?;
+                Some(!self.const_enum_members.contains_key(&symbol_id))
+            })
+            .unwrap_or(true)
+    }
+
+    fn get_constant_value_for_reference_id(
+        &self,
+        reference_id: ReferenceId,
+    ) -> Option<ConstantValue<'a>> {
+        let reference = self.scoping.references.get(reference_id)?;
+        let symbol_id = reference.symbol_id()?;
+        self.const_enum_members.get(&symbol_id).map(|target| target.value.clone())
+    }
+}
+
+impl<'a> MayHaveSideEffectsContext<'a> for ConstantEnumCtx<'_, 'a> {
+    fn annotations(&self) -> bool {
+        true
+    }
+
+    fn manual_pure_functions(&self, _callee: &Expression) -> bool {
+        false
+    }
+
+    fn property_read_side_effects(&self) -> PropertyReadSideEffects {
+        PropertyReadSideEffects::All
+    }
+
+    fn unknown_global_side_effects(&self) -> bool {
+        true
+    }
+}
+
+impl<'a> ConstantEvaluationCtx<'a> for ConstantEnumCtx<'_, 'a> {
+    fn ast(&self) -> oxc_ast::AstBuilder<'a> {
+        self.builder
+    }
+}
+
+/// Process a const enum declaration and evaluate its members
+pub fn process_const_enum(
+    enum_declaration: &TSEnumDeclaration<'_>,
+    scoping: &Scoping,
+    const_enum_table: &mut ConstEnumTable,
+) {
+    let symbol_id = enum_declaration.id.symbol_id();
+    let current_scope = enum_declaration.scope_id();
+    let allocator = oxc_allocator::Allocator::default();
+    let ast_builder = AstBuilder::new(&allocator);
```

Comment:
Use a temp AstBuilder to satisfy `ConstantEvaluationCtx` https://github.com/oxc-project/oxc/blob/0653e9f018468130abc4bf742699e95a23d1daec/crates/oxc_semantic/src/const_enum.rs?plain=1#L188-L192, since the lifetime did not live long enough, we need to convert the arena-based `ConstValue` into a heap-allocated version finally.